### ISSUE TITLE
Can you build backend services using the Functional Programming Paradigm?

### ISSUE THREAD

**Issue by bushidocodes** [bushidocodes/ama#5]
Broken out from https://github.com/bushidocodes/ama/issues/4

Can you actively develop on the backend with a Functional Paradigm? I understand the tenants of FP, but don't get how you apply it server side. I do see how it happens on the front with React, but how do you build DB and map it to your App without some kind of OOP paradigm? Does FP on the back just mean not mutating existing values, but replacing them with values derived from pure functions? It's hard to get what FP means in real deployment situations.

**Comment by bushidocodes** [comment]
I'd say that the backend is actually more suitable for functional programming. Lisp, Scheme, Clojure, Erlang, OCaml, Haskell, F#, and Scala are all backend languages with an FP focus. Compile-to-JS languages like Elm and Reason try to embrace FP, but it seems fundamentally more difficult for these languages to be functional due to the DOM.

In any case, I think that your question is mostly something along the lines of "how do you work with state in FP?" I'd say that this is probably a bit over my head and varies between the languages that I mentioned above, but generally, functional programming isn't "state-less" per-say, but it tries to avoid state mutations and "side effects."

In Redux for example, a store is more-or-less a snapshot of state at a current point in time. An "action" fires a "reducer" that takes the existing store, applies the data in the action, and creates a new store object. The old store and the new store are different objects entirely. This is more of a functional style for applying state changes without mutation. You might want to read up on [Haskell](https://en.wikibooks.org/wiki/Haskell/Understanding_monads/State) if you want to understand about how a backend language does this.

BTW, you might be interested to know that many functional paradigm tools like React actually might use object orientation under the hood. I recently created went through a tutorial series along the lines of "Build your own simplified React 15.x," and it was pretty much classes and instances all the ways down.

**Comment by smlarkin** [comment]
Yeah, I guess you are right.  I am talking about how you deal with state inside a DB.  One of the arguments I've heard for not mutating an array is the cost in swapping bytes is huge.  For instance, it's easer to push / pop than shift / unshift because you have to tweak the same memory allocation instead of just writing out a whole new one.  So FP takes care of this kind of thing by default.

On the other hand, writing pure functions that aim to use zero loops is kind of impossible.  .Map and .Filter and .Reduce have loops built into them.  I understand you are being declarative and everything, but you can't escape using loops.  And I would say that you are going to be busy writing a hell of a lot of functions if you want all to be pure and work inside of single a responsibility ethos.

I guess Haskell is the default functional language for newbies.  I'll have to look into it ... AFTER I feel good about building stuff and have a job :)
